### PR TITLE
Fixes the splitByLine for Polygon with Holes

### PR DESCRIPTION
**Pull request by amir-ba**
### Description:

This PR fixes the issue when using the `SplitByLine` tool for polygons with holes. 

Plz review

### CODE REVIEW

**Comment by amir-ba** [comment]
@marcjansen : This is principally what you proposed. All this time I was adding the drawn polygon from the cut process to the final cut polygons array and this was causing random issues. I had to change it so it takes the intersected polygon part as the result and done.

**Comment by marcjansen** [comment]
Thanks for this @amir-ba, I am going to look into this as soon as the build is green.

**Comment by amir-ba** [comment]
@ahennr thanks for reminding me about the Multipolygons. I added the support for MultiPolygons as well. Please review.

File: src/Util/GeometryUtil/GeometryUtil.js [hunk 1]
```diff
@@ -62,58 +58,77 @@ class GeometryUtil {
...
+    // the array containing all the splited features 
+    let allSplitedPolygons = [];
+    // iterates over each polygon and splits it
+    geometries.forEach(geometry => {
+      // Convert the polygon to turf.js/GeoJSON geometry while
+      // reprojecting them to the internal turf.js projection 'EPSG:4326'.
+      const turfPolygon = geoJsonFormat.writeGeometryObject(geometry);
+      const turfPolygonCoordinates = turfPolygon.coordinates;
+      // ouline lines of the given polygon
+      const outer = lineString(turfPolygonCoordinates[0]);
```
Comment (dnlkoch) [inline src/Util/GeometryUtil/GeometryUtil.js]: Can we surely assume, the first index contains the outer polygon at all times? Maybe we could enhance this by checking the draw direction of the linestring with `booleanClockwise()`.

File: src/Util/GeometryUtil/GeometryUtil.js [hunk 2]
```diff
@@ -62,58 +58,77 @@ class GeometryUtil {
...
+      // polygonized outer polygon
+      const outerPolygon = lineToPolygon(outer);
+      // holes in the polygon
+      const inners = [];
+      turfPolygonCoordinates.slice(1, turfPolygonCoordinates.length).forEach(function (coord) {
+        inners.push(lineString(coord));
+      });
```
Comment (dnlkoch) [inline src/Util/GeometryUtil/GeometryUtil.js]: See comment above.

File: src/Util/GeometryUtil/GeometryUtil.js [hunk 3]
```diff
@@ -62,58 +58,77 @@ class GeometryUtil {
...
+    const geometries = GeometryUtil.separateGeometries(polygonFeat.getGeometry());
+    // the array containing all the splited features 
+    let allSplitedPolygons = [];
+    // iterates over each polygon and splits it
+    geometries.forEach(geometry => {
+      // Convert the polygon to turf.js/GeoJSON geometry while
+      // reprojecting them to the internal turf.js projection 'EPSG:4326'.
+      const turfPolygon = geoJsonFormat.writeGeometryObject(geometry);
+      const turfPolygonCoordinates = turfPolygon.coordinates;
+      // ouline lines of the given polygon
```
Comment (dnlkoch) [inline src/Util/GeometryUtil/GeometryUtil.js]: `ouline` -> `outer`.

File: src/Util/GeometryUtil/GeometryUtil.js [hunk 4]
```diff
@@ -62,58 +58,77 @@ class GeometryUtil {
...
+      turfPolygonCoordinates.slice(1, turfPolygonCoordinates.length).forEach(function (coord) {
+        inners.push(lineString(coord));
+      });
+      // Polygonize the holes in the polygon 
+      const innerPolygon = polygonize(featureCollection(inners));
+      // make a lineString from the spliting line and the outline of the polygon
+      let unionGeom = union(outer, turfLine);
+      // Polygonize the unioned lines.
+      const polygonizedUnionGeom = polygonize(unionGeom);
+      // Array of the splited polygons within the geometry
```
Comment (dnlkoch) [inline src/Util/GeometryUtil/GeometryUtil.js]: `splited` -> `split`.

File: src/Util/GeometryUtil/GeometryUtil.js [hunk 5]
```diff
@@ -62,58 +58,77 @@ class GeometryUtil {
...
-            getCoords(currentSegment)[0],
-            getCoords(currentSegment)[1]
-          );
-        } else {
-          filteredSegments[multiFeatureIndex].push(
-            getCoords(currentSegment)[1]
-          );
+    // This lists all the polygons in the feature and splits the Multi polygons into an array of polygons.
+    const geometries = GeometryUtil.separateGeometries(polygonFeat.getGeometry());
+    // the array containing all the splited features
```
Comment (dnlkoch) [inline src/Util/GeometryUtil/GeometryUtil.js]: `splited` -> `split`.

File: src/Util/GeometryUtil/GeometryUtil.js [hunk 6]
```diff
@@ -62,58 +58,77 @@ class GeometryUtil {
...
+      });
+      // Polygonize the holes in the polygon 
+      const innerPolygon = polygonize(featureCollection(inners));
+      // make a lineString from the spliting line and the outline of the polygon
+      let unionGeom = union(outer, turfLine);
+      // Polygonize the unioned lines.
+      const polygonizedUnionGeom = polygonize(unionGeom);
+      // Array of the splited polygons within the geometry
+      const splitedPolygons = [];
+      // Iterate over each feature in the unined feature and remove sections that are outside the initial polygon and
```
Comment (dnlkoch) [inline src/Util/GeometryUtil/GeometryUtil.js]: `unined` -> `combined`.

File: src/Util/GeometryUtil/GeometryUtil.js (hunk 1 above)
Comment (amir-ba) [reply src/Util/GeometryUtil/GeometryUtil.js]: In the <a href="http://wiki.geojson.org/GeoJSON_draft_version_6">GeoJSON specifications</a>  it states that:
"For type "Polygon", each element in the coordinates array is a coordinates array as described for type "LineString". Furthermore, each LineString in the coordinates array must be a LinearRing. For Polygons with multiple LinearRings, the <b>first must be the exterior ring and any others must be interior rings or holes</b>." 
So I think it is not necessary.

Comment (dnlkoch) [reply src/Util/GeometryUtil/GeometryUtil.js]: Top!

**Pull request merged**